src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1514,6 +1520,9 @@ int tryAllocateThread(final float growthResistance) {
             oldSize = currentSizeOf(oldStat);
             if (oldSize >= maxSizeOf(oldStat)) {
                 // max threads already reached
+                if (reachedMax.compareAndSet(false, true)) {
+                    ThreadDumpUtil.threadDump(maxSizeOf(oldStat));
```

Comment:
What should be the condition?  This has a boolean to only log on the first occurrence of exhaustion.  But if we'd accept such a thread dump utility, I think there should be some further configuration required to enable it for the pool if desired.  So this would then prompt another change primarily in XNIO to enable this thread dumping for its task pool where such dumping would be desired.  Such a dump would likely not be desired from the MSC thread pool upon its exhaustion for instance.

While the behavior of queuing upon is exhaustion is the expected behavior, the reality for customers using JBoss and support engineers assisting them is that this is universally a critical issue state for the XNIO task thread pool used by undertow as that begins to immediately degrade response times if not lead to full request unresponsiveness as well if already processing requests are not quickly completing.  We often see multiple cases a day requiring diagnosis of such thread exhaustion induced by many varied issues and most often not from a conventional deadlock.  Further details about this is recorded at https://issues.redhat.com/browse/EAPSUP-1325.

Reply:
The condition as you suggest should be a configuration switch, and I'd also propose that there should be either two switches or a single three-state switch to separately configure whether a warning is logged when the thread pool queues, and separately whether the warning includes the thread dump.

Reply:
Improved this to make it configurable.  Let me know any suggestions.